website/server/controllers/api-v3/groups.js

```diff
@@ -1005,10 +1002,28 @@ api.removeGroupMember = {
     const message = req.query.message || req.body.message;
     _sendMessageToRemoved(group, member, message, isInGroup);
 
-    await Promise.all([
-      member.save(),
-      group.save(),
-    ]);
+    const session = await mongoose.startSession();
+
+    try {
+      await session.withTransaction(async () => {
+        await member.save({ session });
+        await group.save({ session });
+      }, {
+        retryWrites: true,
+      });
+    } catch (err) {
+      if (err.name === 'MongoError') {
+        throw err.hasErrorLabel('TransactionTooLargeForCache')
+          ? new TransactionError(`Transaction too large for cache: ${err.message}`)
+          : new DatabaseError(`Database error: ${err.message}`);
+      } else if (err.name === 'ValidationError') {
+        throw validationErrors;
+      } else {
+        throw new InternalServerError(`Unexpected error: ${err.message}`);
+      }
+    } finally {
+      session.endSession();
+    }
```

Comment:
```suggestion
    const originalMemberCount = group.memberCount;
    try {
      await Group.db.transaction(async session => {
        group.memberCount -= 1;
        await member.save({ session });
        await group.save({ session });
      }, {
        retryWrites: true,
      });
    } catch (err) {
      group.memberCount = originalMemberCount;
      await group.save();
      if (err.name === 'MongoError') {
        throw err.hasErrorLabel('TransactionTooLargeForCache')
          ? new TransactionError(`Transaction too large for cache: ${err.message}`)
          : new DatabaseError(`Database error: ${err.message}`);
      } else if (err.name === 'ValidationError') {
        throw validationErrors;
      } else {
        throw new InternalServerError(`Unexpected error: ${err.message}`);
      }
    }
```

When not import Mongoose with `startSession`, we could also consider an approach like this.